Blog post about contract upgrades (#7792)
unlock-protocol/unlock: unlock-protocol-com/blog/upgradeable-patterns.md

---
title: "Making locks upgradeable"
authorName: Clément Renaud
publishDate: Nov 18, 2021

> description: "What are contract upgrades on Ethereum, and how we use, test and deploy them at Unlock"

Reviewer: ```suggestion
description: "What are contract upgrades on Ethereum, and how we use, test and deploy them at Unlock"
image: /static/images/blog/unlocking-smart-contracts/code.jpeg
```

  ...
  impl             	Get the contract implementation address
  ...
  upgrade          	Upgrade an existing contract with a new implementation
  upgrade:prepare  	Deploy the implementation of an upgreadable contract
  upgrade:propose  	Send an upgrade implementation proposal to multisig
  ...
  verify           	Verifies contract on Etherscan
  ...
```

## Upgrading contract from another contract

Besides making our main contracts upgradeable, we started to add the upgrade logic inside our product itself. When you deploy a lock with Unlock, a new instance of the `PublicLock` contract (ERC-721) is created with the parameters you defined (name, token, etc). In Solidity, several patterns exist to [deploy a contract from another contract](https://github.com/clemsos/sol-upgrade-pattern/). The previous version of Unlock was using a *Minimal Proxy* (EIP-1167). Instead of deploying the entire contract for each new lock created, a small proxy is deployed just to hold the data and forward calls to a main instance that deals with all the incoming. That allowed to save a non-negligeable amount of gas when deploying a new lock instance.

> However, the minimal proxy approach does not allow for upgrades. We wanted all locks to benefit from the latest features available so we started to investigate how we could deploy a full proxy from a contract and allow users to upgrade their locks when new features were released. The problem were twofold: 1) allow all managers of a lock to upgrade themselves (keep things decentralized), 2) make sure upgrades are safe.

Reviewer: ```suggestion
However, the minimal proxy approach _did not allow for upgrades_. We wanted all locks to benefit from the latest features available so we started to investigate how we could deploy a full proxy from a contract and **allow users to upgrade their locks** when new features were released. The problem was twofold: 1) allow all managers of a lock to upgrade themselves (keep things decentralized), 2) make sure upgrades are safe.
```

```solidity
/**
   * @dev Upgrade a Lock template implementation
   * @param lockAddress the address of the lock to be upgraded
   * @param version the version number of the template
   */
  function upgradeLock(address payable lockAddress, uint16 version) public returns(address) {
    require(proxyAdminAddress != address(0), "proxyAdmin is not set");

    // check perms
    require(_isLockManager(lockAddress, msg.sender) == true, "caller is not a manager of this lock");

    // check version
    IPublicLock lock = IPublicLock(lockAddress);
    uint16 currentVersion = lock.publicLockVersion();
    require( version == currentVersion + 1, 'version error: only +1 increments are allowed');

    // make our upgrade
    address impl = _publicLockImpls[version];
    TransparentUpgradeableProxy proxy = TransparentUpgradeableProxy(lockAddress);
    proxyAdmin.upgrade(proxy, impl);

    emit LockUpgraded(lockAddress, version);
    return lockAddress;
  }
```

> The various versions are stored in the contract with a version number and its corresponding implementation. For instance, lock managers can launch an upgrade by calling `upgradeLock(0x...lockAddress, 10)` to upgrade their lock to version 10 - or though the dashboard. For better compatibility, we forbid bumps of more than one version number (so version 9 to 10 works, but 8 to 10 will fail).

Reviewer: ```suggestion
The various versions are stored in the contract with a version number and its corresponding implementation. For instance, lock managers can launch an upgrade by calling `upgradeLock(0x...lockAddress, 10)` to upgrade their lock to version 10 - or through the dashboard. For better compatibility, we forbid bumps of more than one version number (so version 9 to 10 works, but 8 to 10 will fail).
```
